homeassistant/components/overkiz/executor.py

```diff
@@ -0,0 +1,132 @@
+"""Class for helpers and communication with the OverKiz API."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+from urllib.parse import urlparse
+
+from pyoverkiz.models import Command, Device
+
+from .coordinator import OverkizDataUpdateCoordinator
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class OverkizExecutor:
+    """Representation of an Overkiz device with execution handler."""
+
+    def __init__(
+        self, device_url: str, coordinator: OverkizDataUpdateCoordinator
+    ) -> None:
+        """Initialize the executor."""
+        self.device_url = device_url
+        self.coordinator = coordinator
+        self.base_device_url = self.device_url.split("#")[0]
+
+    @property
+    def device(self) -> Device:
+        """Return Overkiz device linked to this entity."""
+        return self.coordinator.data[self.device_url]
+
+    def select_command(self, *commands: str) -> str | None:
+        """Select first existing command in a list of commands."""
+        existing_commands = self.device.definition.commands
+        return next((c for c in commands if c in existing_commands), None)
+
+    def has_command(self, *commands: str) -> bool:
+        """Return True if a command exists in a list of commands."""
+        return self.select_command(*commands) is not None
+
+    def select_state(self, *states) -> str | None:
+        """Select first existing active state in a list of states."""
+        for state in states:
+            if current_state := self.device.states[state]:
+                return current_state.value
+
+        return None
+
+    def has_state(self, *states: str) -> bool:
+        """Return True if a state exists in self."""
+        return self.select_state(*states) is not None
+
+    def select_attribute(self, *attributes) -> str | None:
+        """Select first existing active state in a list of states."""
+        for attribute in attributes:
+            if current_attribute := self.device.attributes[attribute]:
+                return current_attribute.value
+
+        return None
+
+    async def async_execute_command(self, command_name: str, *args: Any):
+        """Execute device command in async context."""
+        try:
+            exec_id = await self.coordinator.client.execute_command(
+                self.device.device_url,
+                Command(command_name, list(args)),
+                "Home Assistant",
+            )
+        except Exception as exception:  # pylint: disable=broad-except
```

Comment:
Our current standard is to raise an exception, eg HomeAssistantError, with a message as argument in the service handler when a service call fails. The message will be shown in the frontend.